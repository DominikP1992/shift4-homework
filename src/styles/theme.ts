import { defaultTheme } from '@xstyled/emotion';

const theme = {
  ...defaultTheme,
  colors: {
    ...defaultTheme.colors,
    primary: '#423C66',
    'primary-hover': '#635D93',
    'primary-pressed': '#231E47',
    secondary: '#FFF',
    'secondary--hover': '#F7F6FE',
    'secondary--pressed': '#ECE9FC',
    'input-border': '#E9EEF2',
    'input-border--active': '#423C66',
    'font-color': '#423C66',
    'midnight-gray': '#4D6475',
    'purple-gray': '#595D7B',
    'blue-gray': '#1E2A32',
    stroke: '#F4F8FA',
    salmon: '#FFDBCB',
    'icon-button--hover': '#F3F5FE',
    'icon-button--pressed': '#E8EAF2',
    gray: '#EEEEEE',
  },
  sizes: {
    ...defaultTheme.sizes,
    xl: '52px',
  },
  texts: {
    ...defaultTheme.texts,
    xs: {
      fontSize: '12px',
      lineHeight: '17px',
    },
    sm: {
      fontSize: '14px',
      lineHeight: '18px',
    },
    default: {
      fontSize: '16px',
      lineHeight: '20px',
    },
    lg: {
      fontSize: '20px',
      lineHeight: '24px',
    },
    xl: {
      fontSize: '24px',
      lineHeight: '28px',
    },
    '2xl': {
      fontSize: '32px',
      lineHeight: '38px',
    },
  },
};

export default theme;
