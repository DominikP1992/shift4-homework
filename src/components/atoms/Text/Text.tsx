import React from 'react';
import styled from '@emotion/styled';
import theme from '@/styles/theme';

type TextTag = 'p' | 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6' | 'span';

type Font = 'Inter' | 'Work sans' | 'Rubik';

type TextSize = keyof Pick<
  typeof theme.texts,
  'xs' | 'sm' | 'default' | 'lg' | 'xl' | '2xl'
>;

type FontWeight = keyof typeof theme.fontWeights;

const textColorsMap = {
  midnightGray: theme.colors['midnight-gray'],
  purpleGray: theme.colors['purple-gray'],
  blueGray: theme.colors['blue-gray'],
  default: theme.colors.primary,
};

type TextColor = keyof typeof textColorsMap;

interface TextProps extends React.HTMLAttributes<HTMLElement> {
  as?: TextTag;
  size?: TextSize;
  font?: Font;
  color?: TextColor;
  weight?: FontWeight;
}

const DynamicText = styled(
  ({ as = 'p', size, font, color, ...otherProps }: TextProps) =>
    React.createElement(as, otherProps),
)<TextProps>`
  ${({
    size = 'default',
    font = 'Work sans',
    color = 'default',
    weight,
  }: TextProps) => `
    font-size: ${theme.texts[size].fontSize};
    line-height: ${theme.texts[size].lineHeight};
    font-family: ${font};
    color: ${textColorsMap[color]};
    font-weight: ${weight ? theme.fontWeights[weight] : 'inherit'};
  `}
`;

const Text: React.FC<TextProps> = ({ as = 'p', ...rest }) => {
  return <DynamicText as={as} {...rest} />;
};

export default Text;
