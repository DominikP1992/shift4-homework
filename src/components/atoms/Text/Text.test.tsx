import React from 'react';
import Text from './Text'; // Adjust the import path as necessary
import theme from '@/styles/theme'; // Adjust the import path as necessary
import { render, screen } from '@/tests/testsUtils';

describe('Text', () => {
  it('renders correctly with default props', () => {
    const { container } = render(<Text>Sample Text</Text>);
    expect(container.firstChild).toMatchSnapshot();
    expect(container.firstChild).toHaveStyle('font-family: Work sans');
    expect(container.firstChild).toHaveStyle(
      `font-size: ${theme.texts.default.fontSize}`,
    );
    expect(container.firstChild).toHaveStyle(
      `line-height: ${theme.texts.default.lineHeight}`,
    );
    expect(container.firstChild).toHaveStyle(`color: ${theme.colors.primary}`);
  });

  it('applies custom size, font, color, and weight', () => {
    const { container } = render(
      <Text as="h1" size="xl" font="Rubik" color="midnightGray" weight="bold">
        Custom Text
      </Text>,
    );
    expect(container.firstChild).toHaveStyle(
      `font-size: ${theme.texts.xl.fontSize}`,
    );
    expect(container.firstChild).toHaveStyle(
      `line-height: ${theme.texts.xl.lineHeight}`,
    );
    expect(container.firstChild).toHaveStyle('font-family: Rubik');
    expect(container.firstChild).toHaveStyle(
      `color: ${theme.colors['midnight-gray']}`,
    );
    expect(container.firstChild).toHaveStyle(
      `font-weight: ${theme.fontWeights.bold}`,
    );
  });

  it('renders the correct HTML tag', () => {
    render(<Text as="h2">Heading 2</Text>);
    const headingElement = screen.getByText('Heading 2');
    expect(headingElement.tagName).toBe('H2');
  });
});
