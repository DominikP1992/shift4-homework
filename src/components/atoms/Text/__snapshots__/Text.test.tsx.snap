// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text renders correctly with default props 1`] = `
<p
  class="css-lxt2yb"
>
  Sample Text
</p>
`;
